### PR TITLE
Install UncaughtExceptionHandler

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

**Comment by yschimke** [comment]
Example when disabling the ExceptionCatchingExecutor, causes a hang.

image

**Comment by yschimke** [comment]
Better example

Before PR

image

With PR

image

File: okhttp-testing-support/src/main/java/okhttp3/testing/UncaughtExceptionHandlerListener.kt [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ */
+package okhttp3.testing
+
+import org.junit.runner.Description
+import java.io.PrintWriter
+import java.io.StringWriter
+import java.lang.Thread.UncaughtExceptionHandler
+
+/**
+ * installs an aggressive default [UncaughtExceptionHandler] similar to the one found on Android.
```
Comment (swankjesse) [inline okhttp-testing-support/src/main/java/okhttp3/testing/UncaughtExceptionHandlerListener.kt]: Nit: https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

**Comment by yschimke** [comment]
> Let's scope the uncaught exception handler to the test that is currently running? And maybe make our rule a factory for threads that have this thing configured?

My first thought was to configure the thread pool used by Dispatcher, implement afterExecute.  But that seems like tests won't run with the normal code.  And any tests overriding the dispatcher would break it.

So this approach was deliberate to mimic the existing listener. But with parallel runs, it's hard to be sure we catch the right tests.

Can you expand on what you mean by scope to the test? Scoping to the JVM (this approach) or a client/dispatcher are the obvious options.

**Comment by yschimke** [comment]
We also don't want or expect this to fire. 

We will tighten up this logic if we implement https://github.com/square/okhttp/issues/5151

**Comment by yschimke** [comment]
ping @swankjesse 

What are your thoughts on upgrading to junit 5, I think there are cleaner ways to do this if so.

**Comment by swankjesse** [comment]
I think we can do better! Your recent PRs surface a bunch of problems related to threads, unexpected exceptions, and unexpected interruptions.

Why don’t we design an integrated solution? Maybe something that works by wrapping `Runnables`? In tests we know which test caused which `Runnable` to be enqueued, and we could track this.

**Comment by yschimke** [comment]
Will rework.

**Comment by swankjesse** [comment]
I was thinking we could create an ExecutiveService factory or wrapper to manage all of our threads with the same care that we manage sockets.

**Pull request closed without merging**